File: ts/step5-value-object/rawCar.ts
import { CarName } from './carName';
import { Position } from './position';

export class Car {
    private readonly name: CarName;
    private position: Position;

    private static readonly MOVE_THRESHOLD = 4;

    constructor(name: CarName) {
        this.name = name;
        this.position = new Position(0);
    }

    public move(randomValue: number): void {
        if (randomValue >= Car.MOVE_THRESHOLD) {
            this.position = this.position.move();
        }
    }

    public getName(): CarName {
        return this.name;
    }

    public getPosition(): Position {
        return this.position;
    }
}

File: java/step5-value-object/src/main/java/badcode/RawCar.java
package badcode;

import badcode.value_object.CarName;
import badcode.value_object.Position;

public class RawCar {
    private CarName name;
    private Position position;

    public RawCar(String name) {
        this.name = CarName.of(name);
        this.position = Position.of(0);
    }

    public void move(int randomValue) {
        if (randomValue >= 4) {
            this.position = position.increase();
        }
    }

    public String getName() {
        return this.name.getName();
    }

    public int getPosition() {
        return this.position.getPosition();
    }
}
